runtime/vm/CRIUHelpers.cpp

```diff
@@ -1112,57 +1106,65 @@ getNativeString(J9VMThread *currentThread, j9object_t javaString, char **nativeS
+ *     previously suspended other threads stay the same
+ *     (incoming parameters: suspend is FALSE, toggleDelayedHaltThreads is TRUE)
+ * Step 4: resume only those threads NOT marked with the delayed halt flag
+ *     previously resumed other threads stay the same
+ *     now all previously suspended threads are resumed
+ *     (incoming parameters: suspend is FALSE, toggleDelayedHaltThreads is FALSE)
+ * This only affects threads that threadCanRunJavaCode() returns TRUE.
+ * No impact on the current thread.
  *
- * @return true if the thread should be suspended or resumed, false if not
- */
-static bool
-shouldToggleJavaThread(J9VMThread *currentThread, BOOLEAN toggleDebugThreads)
-{
-	J9JavaVM *vm = currentThread->javaVM;
-	bool result = true;
-	if (J9_ARE_ALL_BITS_SET(vm->checkpointState.flags, J9VM_CRIU_IS_JDWP_ENABLED)) {
-		char *threadName = getOMRVMThreadName(currentThread->omrVMThread);
-		releaseOMRVMThreadName(currentThread->omrVMThread);
-		/* all threads started by JDWP begin with "JDWP" in their name */
-		bool isJdwpThread = 0 == strncmp("JDWP", threadName, 4);
-		result = (toggleDebugThreads) ? isJdwpThread : !isJdwpThread;
-	}
-	return result;
-}
-
-/**
- * Suspends or resumes Java threads for checkpoint and restore.
  * Caller must first acquire exclusive safepoint or exclusive VMAccess.
  *
  * @param[in] suspend set to true to suspend threads, and false to resume them
- * @param[in] toggleDebugThreads only useful if JDWP is enabled, determines whether to toggle
- * the debug threads or all other threads
+ * @param[in] toggleDelayedHaltThreads whether to toggle threads marked with the delay halt flag or the threads without that flag
  */
 static void
-toggleSuspendOnJavaThreads(J9VMThread *currentThread, BOOLEAN suspend, BOOLEAN toggleDebugThreads)
+toggleSuspendOnJavaThreads(J9VMThread *currentThread, bool suspend, bool toggleDelayedHaltThreads)
 {
 	J9JavaVM *vm = currentThread->javaVM;
 
 	Assert_VM_criu_true((J9_XACCESS_EXCLUSIVE == vm->exclusiveAccessState) || (J9_XACCESS_EXCLUSIVE == vm->safePointState));
-
+	if (TrcEnabled_Trc_VM_criu_toggleSuspendOnJavaThreads_start) {
+		char *currentThreadName = getOMRVMThreadName(currentThread->omrVMThread);
+		Trc_VM_criu_toggleSuspendOnJavaThreads_start(currentThread, currentThreadName, suspend, toggleDelayedHaltThreads);
+		releaseOMRVMThreadName(currentThread->omrVMThread);
+	}
 	J9VMThread *walkThread = J9_LINKED_LIST_START_DO(vm->mainThread);
 	while (NULL != walkThread) {
 		if (VM_VMHelpers::threadCanRunJavaCode(walkThread)
-		&& (currentThread != walkThread)
+			&& (currentThread != walkThread)
 		) {
-			if (shouldToggleJavaThread(walkThread, toggleDebugThreads)) {
+			bool isDelayHaltFlagSet = J9_ARE_ANY_BITS_SET(walkThread->privateFlags2, J9_PRIVATE_FLAGS2_DELAY_HALT_FOR_CHECKPOINT);
+			/* suspend/resume when (toggleDelayedHaltThreads && isDelayHaltFlagSet) or (!toggleDelayedHaltThreads && !isDelayHaltFlagSet) */
```

Comment:
I don't think this comment helps much: it's just as much work to understand it as the code that was here before. We could say "toggleDelayedHaltThreads and isDelayHaltFlagSet are both true or both false", but that isn't much better.

Perhaps replacing the pair of `bool` parameters with a single parameter that has the four possible values listed in the documentation above would help.

Reply:
Replaced the two bool parameters with
```
#define SUSPEND_THEADS_NO_DELAY_HALT_FLAG 0
#define SUSPEND_THEADS_WITH_DELAY_HALT_FLAG 1
#define RESUME_THREADS_WITH_DELAY_HALT_FLAG 2
#define RESUME_THREADS_NO_DELAY_HALT_FLAG 3
```